File: test/cc/redpen/intellij/SettingsManagerTest.java
package cc.redpen.intellij;

import cc.redpen.config.Configuration;
import org.junit.Before;
import org.junit.Test;

import javax.swing.*;

import static org.mockito.Mockito.*;

public class SettingsManagerTest extends BaseTest {
  SettingsManager manager = spy(new SettingsManager());

  @Before
  public void setUp() throws Exception {
    doNothing().when(manager).restartInspections();
    manager.provider = mock(RedPenProvider.class, RETURNS_DEEP_STUBS);
    manager.settingsPane = mock(SettingsPane.class);
    manager.settingsPane.autodetectLanguage = mock(JCheckBox.class);
  }

  @Test
  public void applyConfigSwitch() throws Exception {
    manager.settingsPane.setConfig(mock(Configuration.class));
    manager.apply();
    verify(manager.provider).setActiveConfig(manager.settingsPane.getConfig());
  }

  @Test
  public void applyAutodetectIfNeeded() throws Exception {
    when(manager.settingsPane.autodetectLanguage.isSelected()).thenReturn(false);
    manager.apply();
    verify(manager.provider).setAutodetect(false);
  }

  @Test
  public void doNotApplyAutodetectIfNotNeeded() throws Exception {
    when(manager.settingsPane.autodetectLanguage.isSelected()).thenReturn(true);
    manager.apply();
    verify(manager.provider).setAutodetect(true);
  }

  @Test
  public void applyValidatorsAndSymbols() throws Exception {
    manager.apply();
    verify(manager.settingsPane).save();
    verify(manager).restartInspections();
  }

  @Test
  public void reset() throws Exception {
    manager.reset();
    verify(manager.settingsPane).resetChanges();
  }
}

File: test/cc/redpen/intellij/SettingsManagerTest.kt
package cc.redpen.intellij

import cc.redpen.config.ValidatorConfiguration
import com.intellij.openapi.project.ProjectManager
import com.nhaarman.mockito_kotlin.mock
import com.nhaarman.mockito_kotlin.spy
import com.nhaarman.mockito_kotlin.verify
import com.nhaarman.mockito_kotlin.whenever
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import org.mockito.Mockito.RETURNS_DEEP_STUBS
import org.mockito.Mockito.doNothing
import java.util.*

class SettingsManagerTest : BaseTest() {
    val config = config("en")
    val manager: SettingsManager

    init {
        val projectManager = mock<ProjectManager>(RETURNS_DEEP_STUBS)
        whenever(application.getComponent(ProjectManager::class.java)).thenReturn(projectManager)
        manager = spy(SettingsManager(project))
    }

    @Before
    fun setUp() {
        doNothing().whenever(manager).restartInspections()
        manager.provider = mock(RETURNS_DEEP_STUBS)
        manager.settingsPane = mock()
        whenever(manager.settingsPane.config).thenReturn(config);
    }

    @Test
    fun applyConfigSwitch() {
        manager.apply()
        verify(manager.provider).activeConfig = config;
    }

    @Test
    fun applyValidatorsAndSymbols() {
        manager.apply()
        verify(manager.settingsPane).save()
        verify(manager).restartInspections()
    }

    @Test
    fun reset() {
        manager.reset()
        verify(manager.settingsPane).resetChanges()
    }

    @Test
    fun isNotModified() {
        val configs = manager.provider.configs
        whenever(manager.settingsPane.configs).thenReturn(configs)
        assertFalse(manager.isModified)
        verify(manager.settingsPane).applyChanges()
    }

    @Test
    fun isModified() {
        whenever(manager.settingsPane.configs).thenReturn(hashMapOf())
        assertTrue(manager.isModified)
        verify(manager.settingsPane).applyChanges()
    }

    @Test
    fun isModified_validatorProperty() {
        val config = configWithValidators(listOf(ValidatorConfiguration("blah")))
        val configs = hashMapOf("en" to config.clone())
        configs["en"]!!.validatorConfigs[0].properties["blah"] = "blah";
        whenever(manager.settingsPane.configs).thenReturn(configs)
        whenever(manager.provider.configs).thenReturn(hashMapOf("en" to config))
        assertTrue(manager.isModified)
        verify(manager.settingsPane).applyChanges()
    }
}